### ISSUE TITLE
Pickling inner/nested classes fails

### ISSUE THREAD

**Issue by jakirkham** [uqfoundation/dill#135]
Supposing I have a module like below.

    ### nest.py
    class A(object):
        class B(object):
            pass

    # B = A.B

If I run this, all is well.

    >>> import dill
    >>> from nest import A
    >>> dill.loads(dill.dumps(A))
    nest.A

However, if I run this, everything comes to a halt. For a workaround, one can uncomment the line at the end of the module.

    >>> dill.loads(dill.dumps(A.B))
    ---------------------------------------------------------------------------
    PicklingError                             Traceback (most recent call last)
    <ipython-input-3-34901c70ec52> in <module>()
    ----> 1 dill.loads(dill.dumps(A.B))
    
    /opt/conda/lib/python2.7/site-packages/dill/dill.pyc in dumps(obj, protocol, byref, fmode, recurse)
        190     """pickle an object to a string"""
        191     file = StringIO()
    --> 192     dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
        193     return file.getvalue()
        194 
    
    /opt/conda/lib/python2.7/site-packages/dill/dill.pyc in dump(obj, file, protocol, byref, fmode, recurse)
        184             return
        185     # end hack
    --> 186     pik.dump(obj)
        187     return
        188 
    
    /opt/conda/lib/python2.7/pickle.pyc in dump(self, obj)
        222         if self.proto >= 2:
        223             self.write(PROTO + chr(self.proto))
    --> 224         self.save(obj)
        225         self.write(STOP)
        226 
    
    /opt/conda/lib/python2.7/pickle.pyc in save(self, obj)
        284         f = self.dispatch.get(t)
        285         if f:
    --> 286             f(self, obj) # Call unbound method with explicit self
        287             return
        288 
    
    /opt/conda/lib/python2.7/site-packages/dill/dill.pyc in save_type(pickler, obj)
       1100        #print ("%s\n%s" % (type(obj), obj.__name__))
       1101        #print ("%s\n%s" % (obj.__bases__, obj.__dict__))
    -> 1102         StockPickler.save_global(pickler, obj)
       1103     return
       1104 
    
    /opt/conda/lib/python2.7/pickle.pyc in save_global(self, obj, name, pack)
        746             raise PicklingError(
        747                 "Can't pickle %r: it's not found as %s.%s" %
    --> 748                 (obj, module, name))
        749         else:
        750             if klass is not obj:
    
    PicklingError: Can't pickle <class 'nest.B'>: it's not found as nest.B

**Comment by mmckerns** [comment]
```
>>> class A(object):
...   class B(object):
...     pass
... 
>>> import dill
>>> dill.dumps(A.B)
'\x80\x02cdill.dill\n_create_type\nq\x00(cdill.dill\n_load_type\nq\x01U\x08TypeTypeq\x02\x85q\x03Rq\x04U\x01Bq\x05h\x01U\nObjectTypeq\x06\x85q\x07Rq\x08\x85q\t}q\n(U\n__module__q\x0bU\x08__main__q\x0cU\x07__doc__q\rNutq\x0eRq\x0f.'
>>> dill.loads(_)
<class '__main__.B'>
```

However, when moving the class def for `A` into a file (`nest.py`), I can reproduce your error (with `dill-0.2.5.dev0`):
```
>>> from nest import A
>>> import dill
>>> dill.dumps(A.B)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.5.dev0-py2.7.egg/dill/dill.py", line 197, in dumps
    dump(obj, file, protocol, byref, fmode, recurse)#, strictio)
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.5.dev0-py2.7.egg/dill/dill.py", line 190, in dump
    pik.dump(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/Users/mmckerns/lib/python2.7/site-packages/dill-0.2.5.dev0-py2.7.egg/dill/dill.py", line 1177, in save_type
    StockPickler.save_global(pickler, obj)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/pickle.py", line 748, in save_global
    (obj, module, name))
pickle.PicklingError: Can't pickle <class 'nest.B'>: it's not found as nest.B
```

**Comment by jakirkham** [comment]
Correct running the code in the REPL won't give this error. However, this is a simple reproducer. In reality, a reasonably sized module is in use. So, I would rather have a solution that uses importing.

**Comment by mmckerns** [comment]
What do you mean "uses importing"?

**Comment by jakirkham** [comment]
Yeah, maybe that was the wrong choice of words especially given how I understand this library solves problems in some cases.

Let's put it this way, I don't want to have to explicitly evaluate my code in multiple places to get it to work. However, if your solution is to do this under the hood, I am completely ok with that. Does that make more sense?

**Comment by mmckerns** [comment]
No, that's not any clearer.  The problem is one of namespacing, and it's not a simple one.  It's solved in `python 3` by each object having the `__qualname__` attribute.
```
>>> from nest import A
>>> A.B.__qualname__
'A.B'
>>> import dill
>>> dill.dumps(A.B)
b'\x80\x03cdill.dill\n_get_attr\nq\x00X\x08\x00\x00\x00builtinsq\x01X\x07\x00\x00\x00getattrq\x02\x86q\x03Rq\x04cnest\nA\nq\x05X\x01\x00\x00\x00Bq\x06\x86q\x07Rq\x08.'
>>> dill.loads(_)
<class 'nest.A.B'>
```
Note:
```
>>> sys.version
'3.5.0 (default, Sep 15 2015, 23:57:10) \n[GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)]'
```

**Comment by mmckerns** [comment]
Are you saying, simply, you are looking for a similar solution from `dill` for `python 2`?

**Comment by jakirkham** [comment]
Sure, actually that seems like a fine solution. Thanks.

**Comment by jakirkham** [comment]
Maybe using a dependency like this [`qualname`]( https://pypi.python.org/pypi/qualname ) would solve the problem then?

**Comment by mmckerns** [comment]
@jakirkham: Yes, that is something like what is needed.  It only works for objects defined within a file, however `dill.source` should be able to extend it to work in `__main__`.  I'll have a look at it, and maybe `dill.source` includes most of what `qualname` is doing already.  Thanks for the pointer.

**Comment by mmckerns** [comment]
I just read the source for `qualname`.  I already have most of this inside `dill.source`, and I believe handles more cases than `qualname`… but yes, this is a viable solution, and I think I can take the one element that's in this package and make it more generally apply in `dill`.

**Comment by mmckerns** [comment]
Continuing this thread on the active issue...

**Comment by jakirkham** [comment]
It looks like this got put on the 0.2.5 release. Does that mean there is a fix for issue ( https://github.com/uqfoundation/dill/issues/56 )?

**Comment by mmckerns** [comment]
It's just that the issue was closed (as duplicate) during the 0.2.5 dev period, so it goes into that release.  The original is still open.
